src/dfi/validation.cpp

```diff
@@ -2506,124 +2652,1137 @@ static void ExecuteTokenSplits(const CBlockIndex *pindex,
+            if (!res) {
+                return res;
+            }
+        } else {
+            addView.AddBalance(owner, {poolPair->idTokenA, resAmountA});
+        }
+
+        if (tokensToBeLocked.count(poolPair->idTokenB.v)) {
+            LogPrintf("locking from pool %d@%d\n", resAmountB, poolPair->idTokenB.v);
+            res = LockToken(cache, pindex->nHeight, pindex->GetBlockHash(), owner, {poolPair->idTokenB, resAmountB});
+            if (!res) {
+                return res;
+            }
+        } else {
+            res = addView.AddBalance(owner, {poolPair->idTokenB, resAmountB});
+            if (!res) {
+                return res;
+            }
+        }
+        addView.Flush();
+
+        // TODO: keep affected pools in memory and update all at the end
+        res = cache.SetPoolPair(lpAmount.nTokenId, pindex->nHeight, *poolPair);
+        if (!res) {
+            return res;
+        }
+        cache.SetShare(lpAmount.nTokenId, owner, pindex->nHeight);
+    }
+
+    // from vault collaterals (only USDD)
+    cache.ForEachVaultCollateral([&](const CVaultId &vaultId, const CBalances &balances) {
+        for (const auto &[tokenId, amount] : balances.balances) {
+            if (tokensToBeLocked.count(tokenId.v)) {
+                auto owner = cache.GetVault(vaultId)->ownerAddress;
+
+                const auto amountToLock = lockedAmount(amount);
+
+                LogPrintf("locking from collateral %d@%d\n", amountToLock, tokenId.v);
+                res = cache.SubVaultCollateral(vaultId, {tokenId, amountToLock});
+                if (!res) {
+                    return false;
+                }
+                res = LockToken(cache, pindex->nHeight, pindex->GetBlockHash(), owner, {tokenId, amountToLock});
+                if (!res) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    });
+
+    if (!res) {
+        return res;
+    }
+
+    CDataStructureV0 releaseKey{AttributeTypes::Live, ParamIDs::Economy, EconomyKeys::TokenLockRatio};
+    auto attributes = cache.GetAttributes();
+    attributes->SetValue(releaseKey, lockRatio);
+    cache.SetVariable(*attributes);
+    cache.Flush();
+
+    return res;
+}
+
+static void ProcessTokenLock(const CBlock &block,
+                             const CBlockIndex *pindex,
+                             CCustomCSView &cache,
+                             BlockContext &blockCtx) {
+    const auto &consensus = blockCtx.GetConsensus();
+    if (pindex->nHeight < consensus.DF25Height) {
+        return;
+    }
+
+    const auto attributes = cache.GetAttributes();
+
+    CDataStructureV0 heightKey{AttributeTypes::Param, ParamIDs::dTokenRestart, DFIPKeys::BlockHeight};
+    const auto restartHeight = attributes->GetValue(heightKey, CAmount{});
+
+    if (restartHeight != pindex->nHeight) {
+        return;
+    }
```

Comment:
@Bushstar how about we do it like oracle splits? like that:
`
CDataStructureV0 lockRatioKey{AttributeTypes::Param, ParamIDs::dTokenRestart, static_cast<uint32_t>(pindex->nHeight)};
const auto lockRatio = attributes->GetValue(lockRatioKey, CAmount{});
if(lockRatio == 0) {
  return;
}
`
Then the attribute defines the locked ratio too.